software/apps/baremetal/mimo_mmse_f8/main.c

```diff
@@ -0,0 +1,181 @@
+// Copyright 2021 ETH Zurich and University of Bologna.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+// Author: Marco Bertuletti, ETH Zurich
+
+#include <math.h>
+
+#include "dma.h"
+#include "encoding.h"
+#include "printf.h"
+#include "runtime.h"
+#include "synchronization.h"
+
+#include "baremetal/mempool_checks.h"
+#include "baremetal/mempool_cholesky_f16s.h"
+#include "baremetal/mempool_linearsolver_f16s.h"
+#include "baremetal/mempool_mimo_mmse_f8s.h"
+
+#include "data_mimo_mmse_f8.h"
+#define ZF (0)   // When asserted use zero-forcing
+#define FOLD (0) // When asserted fold matrixes in memory
+#define NUM_BANKS (BANKING_FACTOR * NUM_CORES)
+#define PARALLEL
+#define VEC
+
+#if FOLD
+
+#define NUM_ROW (1 + ((N_ITR * N_TX - 1) / NUM_BANKS))
+__fp16 l1_G[2 * N_TX * NUM_BANKS * NUM_ROW]
+    __attribute__((aligned(sizeof(int32_t)), section(".l1_prio")));
+__fp16 l1_L[2 * N_TX * NUM_BANKS * NUM_ROW]
+    __attribute__((aligned(sizeof(int32_t)), section(".l1_prio")));
+#else
+__fp16 l1_G[2 * N_TX * N_TX * N_ITR]
+    __attribute__((aligned(sizeof(int32_t)), section(".l1_prio")));
+__fp16 l1_L[2 * N_TX * N_TX * N_ITR]
+    __attribute__((aligned(sizeof(int32_t)), section(".l1_prio")));
+#endif
+
+__fp8 l1_H[2 * N_TX * N_RX * N_ITR]
+    __attribute__((aligned(sizeof(int32_t)), section(".l1_prio")));
+__fp8 l1_S[2 * N_TX * N_ITR]
+    __attribute__((aligned(sizeof(int32_t)), section(".l1")));
+__fp8 l1_y[2 * N_RX * N_ITR]
+    __attribute__((aligned(sizeof(int32_t)), section(".l1")));
+__fp16 y2[2 * N_TX * N_ITR]
+    __attribute__((aligned(sizeof(int32_t)), section(".l1")));
+__fp16 y3[2 * N_TX * N_ITR]
+    __attribute__((aligned(sizeof(int32_t)), section(".l1")));
+__fp16 l1_x[2 * N_TX * N_ITR]
+    __attribute__((aligned(sizeof(int32_t)), section(".l1")));
+
+// Driver program
+int main() {
+
+  uint32_t core_id = mempool_get_core_id();
+#ifndef BANSHEE
+  uint32_t num_cores = mempool_get_core_count();
+  mempool_barrier_init(core_id); // Initialize barrier and synchronize
+#endif
```

Comment:
Thank you for noticing, this can be removed here. The special treatment of barriers comes from the work on the end-to-end MIMO to run single-core Monte Carlo simulations with Banshee.